File: Point.py
from __future__ import annotations

from copy import copy
from dataclasses import dataclass

from Vector import Vector

@dataclass
class Point:
    x: int = 0
    y: int = 0
    z: int = 0

    def __post_init__(self) -> None:
        # Ensure components are ints
        self.x = int(self.x)
        self.y = int(self.y)
        self.z = int(self.z)

    def getMidpoint(p: Point, q: Point) -> Point:
        return Point((p.x+q.x)//2, (p.y+q.y)//2, (p.z+q.z)//2)

    def __iadd__(self, other: Vector) -> Point:
        self.x += int(other.x)
        self.y += int(other.y)
        self.z += int(other.z)
        return self
    def __add__(self, other: Vector) -> Point:
        new = copy(self)
        new += other
        return new

    def __isub__(self, other: Vector) -> Point:
        self.x -= int(other.x)
        self.y -= int(other.y)
        self.z -= int(other.z)
        return self
    def __sub__(self, other: Vector) -> Point:
        new = copy(self)
        new -= other
        return new

File: Point.ts
import { Vector } from './Vector.js';

export class Point {
	x = 0;
	y = 0;
	z = 0;

	constructor(x?: number, y?: number, z?: number) {
		this.x = x ?? 0;
		this.y = y ?? 0;
		this.z = z ?? 0;
	}

	getMidpoint(p: Point, q: Point) {
		return new Point(
			(p.x + q.x) / 2,
			(p.y + q.y) / 2,
			(p.z + q.z) / 2
		);
	}

	equals(p: Point) {
		return this.x === p.x && this.y === p.y && this.z === p.z;
	}

	add(vector: Vector) {
		this.x += vector.x;
		this.y += vector.y;
		this.z += vector.z;
	}

	subtract(vector: Vector) {
		this.x -= vector.x;
		this.y -= vector.y;
		this.z -= vector.z;
	}
}
